sdks/java/io/mqtt/src/main/java/org/apache/beam/sdk/io/mqtt/MqttIO.java

```diff
@@ -0,0 +1,492 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io.mqtt;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.auto.value.AutoValue;
+
+import java.io.IOException;
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingQueue;
+
+import javax.annotation.Nullable;
+
+import org.apache.beam.sdk.coders.Coder;
+import org.apache.beam.sdk.coders.SerializableCoder;
+import org.apache.beam.sdk.coders.VoidCoder;
+import org.apache.beam.sdk.io.UnboundedSource;
+import org.apache.beam.sdk.options.PipelineOptions;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.transforms.display.DisplayData;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.PBegin;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.PDone;
+
+import org.eclipse.paho.client.mqttv3.IMqttDeliveryToken;
+import org.eclipse.paho.client.mqttv3.MqttCallback;
+import org.eclipse.paho.client.mqttv3.MqttClient;
+import org.eclipse.paho.client.mqttv3.MqttMessage;
+
+import org.joda.time.Duration;
+import org.joda.time.Instant;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * An unbounded source for MQTT broker.
+ *
+ * <h3>Reading from a MQTT broker</h3>
+ *
+ * <p>MqttIO source returns an unbounded collection of {@code byte[]} as
+ * {@code PCollection<byte[]>}, where {@code byte[]} is the MQTT message payload.</p>
+ *
+ * <p>To configure a MQTT source, you have to provide a MQTT connection configuration including
+ * {@code ClientId}, a {@code ServerURI}, and eventually a {@code Topic} pattern. The following
+ * example illustrates various options for configuring the source:</p>
+ *
+ * <pre>{@code
+ *
+ * pipeline.apply(
+ *   MqttIO.read()
+ *    .withMqttConnectionConfiguration(MqttIO.MqttConnectionConfiguration.create(
+ *      "tcp://host:11883",
+ *      "my_client_id",
+ *      "my_topic"))
+ *
+ * }</pre>
+ *
+ * <h3>Writing to a MQTT broker</h3>
+ *
+ * <p>MqttIO sink supports writing {@code byte[]} to a topic on a MQTT broker.</p>
+ *
+ * <p>To configure a MQTT sink, as for the read, you have to specify a MQTT connection
+ * configuration with {@code ClientId}, {@code ServerURI}, {@code Topic}.</p>
+ *
+ * <p>Eventually, you can also specify the {@code Retained} and {@code QoS} of the MQTT
+ * message.</p>
+ *
+ * <p>For instance:</p>
+ *
+ * <pre>{@code
+ *
+ * pipeline
+ *   .apply(...) // provide PCollection<byte[]>
+ *   .MqttIO.write()
+ *     .withMqttConnectionConfiguration(MqttIO.MqttConnectionConfiguration.create(
```

Comment:
* I don't quite like that the word "mqtt" repeats 3 times in this single line.
Consider doing some renames. E.g. ideally it could look like:
.withConnectionConfiguration(MqttIO.ConnectionConfiguration.create(blah))
though this repeats "ConnectionConfiguration" twice... but it's still better.

* in JDBC, the point of introducing JdbcConnectionParameters common between read/write transform was that you could conceivably pass the same JdbcConnectionParameters object to a read transform and a write transform. Is this also the case for MQTT? E.g. do all 3 parameters (server, client id, topic) make sense both for reader and writer?